spec/index.bs

```diff
@@ -7281,6 +7281,24 @@ Issue: Because timestamp query provides high-resolution GPU timestamp, we need t
 
 # Canvas Rendering &amp; Swap Chains # {#canvas-rendering}
 
+## HTMLCanvasElement.getContext ## {#canvas-getcontext}
+
+A {{GPUCanvasContext}} object can be obtained via the `getContext` method of a {{HTMLCanvasElement}} instance:
+
+Issue: WebIDL does not define syntax to use string literals as argument types, use an enum as a workaround.
```

Comment:
Interestingly, WebGL WebIDL doesn't try to define this. Maybe we shouldn't either?

Reply:
The other side of the argument is that [MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext) lists the `contextType` parameter as a list of string literals, which makes it an `enum` in WebIDL and TypeScript. I assume that MDN doesn't type it as such because `enum` is not a JavaScript construct.

On the plus side, having this declaration in the WebIDL side of things allows for automatic type generation, [see this PR](https://github.com/gpuweb/types/pull/58)

[TypeScript does it even weirder](https://github.com/microsoft/TypeScript/blob/d9f5a85d5b7d1094cbfedd7d4792548db93fe74a/lib/lib.webworker.d.ts#L2473), they declare overloads of the function with string literals as parameter types:
```
    getContext(contextId: "2d", options?: CanvasRenderingContext2DSettings): OffscreenCanvasRenderingContext2D | null;
    getContext(contextId: "bitmaprenderer", options?: ImageBitmapRenderingContextSettings): ImageBitmapRenderingContext | null;
    getContext(contextId: "webgl", options?: WebGLContextAttributes): WebGLRenderingContext | null;
    getContext(contextId: "webgl2", options?: WebGLContextAttributes): WebGL2RenderingContext | null;
    getContext(contextId: OffscreenRenderingContextId, options?: any): OffscreenRenderingContext | null;
```

All in all I think this is a difficult one, on one side we could leave it as a string, add a special rule on the types generation and let MDN and TypeScript deal with it ad-hoc... or we could add this declaration but risk going against previous precedents.

From my point of view, if we don't define these kind of constraints in the spec, why define enums at all?
At the end of the day, once this hits JavaScript land, nothing is enforced and users are free to shoot themselves in the foot, the spec should be there to guide them and explain how to use the API in its entirety. Why would we actively omit a function most users (at least on the graphical side of WebGPU) will use very often?

Reply:
TypeScript defines overloads because they can, and it's more expressive. WebIDL cannot express these as overloads. And since it can't, we can't add a `partial interface` in our spec because it will conflict with the upstream Canvas spec.

`getContext` is defined in the HTML Canvas specification:
https://html.spec.whatwg.org/multipage/canvas.html#dom-canvas-getcontext
which calls (implicitly?) into WebGL:
https://www.khronos.org/registry/webgl/specs/latest/2.0/#2.1

So we should do the same ultimately.

Reply:
Yeah, that's a reasonable argument, we don't want to generate conflicts with upstream specs.

I have changed the PR to include an example of getting a `GPUCanvasContext` from a canvas object instead of adding it to the IDL.

Just to clarify, I thought adding `partial interface` to extend an upstream spec was ok because [it is already being done](https://github.com/gpuweb/gpuweb/blob/main/spec/index.bs#L1179) in this spec:
```webidl
<script type=idl>
[Exposed=Window]
partial interface Navigator {
    [SameObject] readonly attribute GPU gpu;
};
</script>
```
Wouldn't that declaration conflict with the [system state spec](https://html.spec.whatwg.org/multipage/system-state.html#navigator)?

Shouldn't we replace those declarations with examples as well for clarity and consistency?

Reply:
No, it's not the `partial interface` that conflicts, it's specifically the `getContext` definition which conflicts with the [existing one](https://html.spec.whatwg.org/multipage/canvas.html#the-canvas-element): it has the same name, so it has to be an overload, but this is not a valid "distinguishable" overload as WebIDL only has general string types (`DOMString` in this case).

Reply:
Understood, but shouldn't we follow the same pattern established in the system state spec?
```WebIDL
interface mixin NavigatorGPU {
    [SameObject] readonly attribute GPU gpu;
}

Navigator includes NavigatorGPU;
WorkerNavigator includes NavigatorGPU;
```

I can submit that change in a separate PR though

Reply:
Ah. Probably. I think I copied this pattern from WebUSB https://wicg.github.io/webusb/#enumeration
but if there's something that indicates we should do it with a mixin (which makes sense to me), we should.